src/category_theory/sites/grothendieck.lean

```diff
@@ -0,0 +1,268 @@
+/- Author: E.W.Ayers.
+   This section roughly follows Chapter 3, §1, §2 of Sheaves in Geology and Logic by Saunders Maclane and Ieke M.
+ -/
+
+import category_theory.sites.sieves
+
+universes u v w
+namespace category_theory
+
+open category_theory limits order lattice
+
+/-- A set of sieves for every object in the category: a candidate to be a Grothendieck topology. -/
+def sieve_set (C : Type u) [category.{v} C] := Π (X : C), set (sieve X)
+
+def sieve_set.trivial (C : Type u) [category.{v} C] : sieve_set C := λ X, {⊤}
+
+lemma mem_trivial (C : Type u) [category.{v} C] {X : C} (S : sieve X) :
+  S ∈ sieve_set.trivial C X ↔ S = ⊤ :=
+set.mem_singleton_iff
+
+/-- A sieve on `X` is dense if for any arrow `f : Y ⟶ X`, there is a `g : Z ⟶ Y` with `g ≫ f ∈ S`. -/
+def sieve_set.dense (C : Type u) [category.{v} C] : sieve_set C :=
+λ X, {S | ∀ {Y : C} (f : Y ⟶ X), ∃ Z (g : Z ⟶ Y), S.arrows (g ≫ f)}
+
+/-- The atomic sieve_set just contains all of the non-empty sieves. -/
+def sieve_set.atomic (C : Type u) [category.{v} C] : sieve_set C :=
+λ X, {S | ∃ {Y} (f : Y ⟶ X), S.arrows f}
+
+open sieve category
+
+/--
+Definition of a Grothendieck Topology: a set of sieves `J X` on each object `X` satisfying three axioms:
+1. For every object `X`, the maximal sieve is in `J X`.
+2. If `S ∈ J X` then its pullback along any `h : Y ⟶ X` is in `J Y`.
+3. If `S ∈ J X` and `R` is a sieve on `X`, then provided that the pullback of `R` along any arrow
+   `f : Y ⟶ X` in `S` is in `J Y`, we have that `R` itself is in `J X`.
+-/
+class grothendieck {C : Type u} [category.{v} C] (J : sieve_set C) : Prop :=
```

Comment:
```suggestion
class grothendieck_top {C : Type u} [category.{v} C] (J : sieve_set C) : Prop :=
```

Reply:
Is `grothendieck_topology` okay? I'm hesitant to use `top` since it clashes with `top` for partial orders.

Reply:
Yup, you are right. Let's go for `grothendieck_topology`

Reply:
I've made this change - but now that I think about it, `grothendieck_topology` as I defined it is a predicate which is a class; perhaps it would be more sensible for that name to be the bundled version instead? That is, for it to be a structure which contains the set of sieves together with the proofs that it's a Grothendieck topology?  I think most of this file is unbundled but I have no objections to bundling